### ISSUE TITLE
Questions about de-serialization of nested classes and custom record constructors

### ISSUE THREAD

**Issue by stevelr** [mtth/avsc#11]
Thank you for the great avsc library and for your support. I've been stuck on two issues and would like to ask for your kind guidance.

1. I haven’t been able to get the type.fromBuffer() method to return an instance of my class -it always returns a generic class. I’ve even set the _constructor field of the Type object to my class constructor before calling fromBuffer(), and verified with the browser debugger that the function signature is the same as with the dynamically-generated constructor. I'm not sure if this is the reason for Resolver, since my reader and writer classes are the same, but I tried creating a resolver with type.createResolver(type) and passing that in to parse, but it created a different error.

2. The second related issue is that my classes have other embedded classes, and pre-registering them results in thrown exceptions. For example, the record definition has a structure like this:
```
    { name: “Person”,
           type: “record”,
           fields: [ {
                     name: “id”,
                     type: {
                           name: “EmployeeRecord”,
                           type: “record”,
                           fields: [
                                { name: “employeeId”, type: long },
                                { name: “startDate”, type: long }
                           ]
                    } },
                   { name: “name”, type: “string” }
             ]
      }
```

I want to pre-register all classes before calling fromBuffer, because I know they will be needed for deserialization. When registering classes, with either avsc.parse() or avsc.types.Type.fromSchema(), there is an Error thrown in the Type constructor because of duplicates.  If "Person" has EmployeeRecord and "Manager" has EmployeeRecord, you get this exception. One thing I haven't tried is editing the schema definition to use references but not full nested record definitions, but these schemas are automatically generated by apache-avro (for java), so I would have to change that schema-generator code or hand edit the schemas, neither of which seems like the right thing to do. I don't know why registering twice should be a problem - imho it should be idempotent. What I want to do is loop through all my classes at startup time and parse/register all of them, in preparation for receiving complex nested objects over the wire. If instead of throwing an Error in line 72 of types.js, it returns (before adding the type to the registry), I can at least pre-register all my classes. When it returns, I also set the _constructor to the correct class constructor. 
I'm probably missing something here but I think it would be helpful if
(a) either duplicate registration is harmless, or there's a flag in opts that can be set to make it not throw an exception
(b) there's an API to specify the record constructor - a setter to complement getRecordConstructor(). 
(c) another variation on (a)&(b) would be to have a pre-register API function that takes a list of (schema,constructor) pairs and registers them, perhaps returning a registry dictionary mapping schema name to Type. Any duplicate registrations within this set caused by nested record definitions would be ignored.
Thanks again for your help and for avsc!

**Comment by stevelr** [comment]
To test the theory, I implemented a variation of (c) by adding a "constructors"  map to the parse options. constructors contains a map from schema name to constructor function. (usage is below). I noticed there are some other properties of the Record class so I just replicated those so it acts like a mixin:
Output of diff -w types.js:
72c72,73
<     throw new Error(f('duplicate type name: %s', name));
---
>     //throw new Error(f('duplicate type name: %s', name));
>     return;
1357c1358,1359
<   this._constructor = this._createConstructor(schema.type === 'error');
---
>   var ctor = opts.constructors && opts.constructors[this._name];
>   this._constructor = this._createConstructor(schema.type === 'error', ctor);
1365c1367,1368
< RecordType.prototype._createConstructor = function (isError) {
---
> RecordType.prototype._createConstructor = function (isError, ctor) {
>
1366a1370,1373
>   var Record;
>   if (typeof(ctor) === 'function') {
>     Record = ctor;
>   } else {
1392c1399,1400
<   var Record = new Function(outerArgs.join(), outerBody).apply(undefined, ds);
---
>     Record = new Function(outerArgs.join(), outerBody).apply(undefined, ds);
>   }
1396,1403c1404,1411
<   Record.prototype = {
<     $clone: function () { return self.clone(this); },
<     $compare: function (obj) { return self.compare(this, obj); },
<     $getType: Record.getType,
<     $isValid: function (opts) { return self.isValid(this, opts); },
<     $toBuffer: function (noCheck) { return self.toBuffer(this, noCheck); },
<     $toString: function () { return self.toString(this); }
<   };
---
>   Record.prototype = Record.prototype || {};
>   Record.prototype.$clone = function () { return self.clone(this); };
>   Record.prototype.$compare = function (obj) { return self.compare(this, obj); };
>   Record.prototype.$getType = Record.getType;
>   Record.prototype.$isValid = function (opts) { return self.isValid(this, opts); };
>   Record.prototype.$toBuffer = function (noCheck) { return self.toBuffer(this, noCheck); };
>   Record.prototype.$toString = function () { return self.toString(this); };
>


------
Setup code looks roughly like this:
var classes = [ Person, ... ]  
      // list of all my predefined classes that might be avro- encoded or decoded
      // each class defines a static function getClassSchema() that returns the avro schema
var parseOpts = { registry: {}, constructors: {} };
// pre-fill constructors
for (var ix in classes) {
  var schema = classes[ix].getClassSchema();
  parseOpts.constructors[schema.name] = classes[ix];
}
// register the classes
for (var ix in classes) {
  var schema = classes[ix].getClassSchema();
  avsc.parse(schema, parseOpts);
}

// later, when calling type.fromBuffer, the correct class is instantiated!

**Comment by mtth** [comment]
Thanks for the kind words, I'm glad you find this library useful.

1. I'm guessing you can't simply extend the generated constructor since you are using TypeScript to generate classes already (correct me if not). I like the idea of passing an optional constructor when the `RecordType` gets instantiated better than adding a `setRecordConstructor` method (having types immutable makes code much simpler to understand). I'd rather not add a new option to `parse` though since these should ideally not be type-specific, so I'm thinking of adding a third optional argument to `RecordType`'s constructor (`RecordType(schema, opts, recordConstructor)`). You could use it via a `typeHook`, something like:

  ```javascript
  function typeHook(schema, opts) {
    if (schema.type === 'record') {
      // Assuming `constructors` is your map of constructors.
      return new avsc.types.RecordType(schema, opts, constructors[schema.name]);
    }
  }
  ```
  (Also, avoid using "private" properties - prefixed by an underscore, e.g. `_constructor` - since these aren't guaranteed to be compatible from one release to another.)

2. It's better to avoid duplicate name definitions. For example, it's not obvious when types would be considered equal (using the canonical representation is not enough: defaults might be different, in which case it's unclear which to use). There are multiple ways around this though, such as:
  + Registering the types separately (each schema only containing a single definition of each name).
  + Implementing a `typeHook` which skips already registered names.

Hopefully this would address your use-case. Let me know what you think.

**Comment by stevelr** [comment]
TL:DR: I didn't fully understand typeHook before, but this looks like a good approach to solving both issues.

You are correct that the classes are auto-generated. I would prefer to avoid creating an external dependency by making my classes depend on an external base Record class. Conceptually, these classes should be independent of transport encoding. The reason they are auto-generated from avro at all is that for consistency/DRY there should be one definition source, and avro idl seemed to be the most straightforward.  (*)

I agree on the principles of avoiding use of private properties, and designing-in immutability - Those were side effects of hacking to understand the problem.

Regarding duplicate classes, the problem arises because I don't have complete control over when RecordType constructor is called. For nested classes, it's called internally within the avsc library. That means that if I want to make sure all the classes are registered exactly once, I would have to keep track of the dependency tree and register the classes in a very specific order. That would be high maintenance and extremely fragile, and in some cases impossible (such as when record type A embeds type B and record type C embeds type B.).

Making a typeHook that skips duplicates would address this, and would allow more precise control than always-allow (my original suggestion) or always-deny (the current implementation). 

(footnote *): The typescript class is generated with a no-args constructor "MyClass()" that fills in all default values declared in the avro spec. Since javascript doesn't support overloaded constructors, the all-args constructor needed for RecordType is a different static method. So, in my case the value of constructors[schema.name] is a static property of MyClass, not MyClass itself. That's why I had to initialize constructor.prototype={} in the sample code I posted, but it could be done inside the custom typeHook. This is just FYI - I don't believe this affects your proposed solution.

Thanks again for help and suggestions!

**Comment by mtth** [comment]
That makes sense, thanks for the detailed feedback.

`2.5.0`, [just released](https://www.npmjs.com/package/avsc), adds the custom constructor option to [`RecordType`](https://github.com/mtth/avsc/wiki/API#class-recordtypeattrs-opts-recordconstructor). One difference with your diff's implementation is that no convenience methods get attached to the custom constructor's prototype (they can always be added manually, as they are simply proxies to the `type`'s own methods).

Usage is as described above. For your use-case, you would also include the deduplication logic inside the `typeHook` (`2.5.0` also introduces a `getName` method available on all types which should make this easier).

**Comment by mtth** [comment]
FYI, the next major release (`3.0`) will include support for [logical types](https://avro.apache.org/docs/current/spec.html#Logical+Types). These are very powerful and in particular provide a way to override the default record constructor. As an unfortunate consequence, the custom argument added above is now redundant and will be deprecated in future `2.*` releases (and removed in `3.0`).

If/when you'll upgrade, migration will be straightforward. You will need to implement a `LogicalType`, as described below. As a bonus, you will be able to use your static factory methods cleanly (without the superfluous `new`).

```javascript
var util = require('util');

function CustomRecordType(attrs, opts, recordFactory) {
  avsc.types.LogicalType.call(this, attrs, opts);
  this._factory = recordFactory;
}
util.inherits(CustomRecordType, avsc.types.LogicalType);

CustomRecordType.prototype._fromValue = function (val) { return this._factory(val); };
CustomRecordType.prototype._toValue = function (rec) { return rec; };
```
Using it can be done using a `typeHook` similar to the previous one:
```javascript
var factories = { /* All your record factories, keyed by record name. */ };

function typeHook(attrs, opts) {
  if (attrs.type === 'record') {
    return new CustomRecordType(attrs, opts, factories[attrs.name]);
  }
}
```
